packages/adapter-utils/src/utils.ts

```diff
@@ -152,91 +253,17 @@ export const transformValues = async (
-          new Field(
-            field.parent,
-            field.name,
-            fieldType.refInnerType,
-            field.annotations
-          ),
-        ))
-      if (!_.isArray(newVal)) {
-        if (strict) {
-          log.debug(`Array value and isListType mis-match for field - ${field.name}. Got non-array for ListType.`)
-        }
-        return transformListInnerValue(newVal)
-      }
-      const transformed = await awu(newVal)
-        .map((v, i) => transformListInnerValue(v, i))
-        .filter((val: Value) => !_.isUndefined(val))
-        .toArray()
-      return transformed.length === 0 && (newVal.length > 0 || !allowEmpty)
-        ? undefined
-        : transformed
-    }
-    if (_.isArray(newVal)) {
-      // Even fields that are not defined as ListType can have array values
-      const transformed = await awu(newVal)
-        .map((item, index) => transformValue(item, keyPathID?.createNestedID(String(index)), field))
-        .filter(val => !_.isUndefined(val))
-        .toArray()
-      return transformed.length === 0 && (newVal.length > 0 || !allowEmpty)
-        ? undefined
-        : transformed
-    }
-
-    if (isObjectType(fieldType) || isMapType(fieldType)) {
-      if (!_.isPlainObject(newVal)) {
-        if (strict) {
-          log.debug(`Value mis-match for field ${field?.name} - value is not an object`)
-        }
-        // _.isEmpty returns true for primitive values (boolean, number)
-        // but we do not want to omit those, we only want to omit empty
-        // objects, arrays and strings. we don't need to check for objects here
-        // because we cannot get here with an object
-        const valueIsEmpty = (Array.isArray(newVal) || _.isString(newVal)) && _.isEmpty(newVal)
-        return (valueIsEmpty && !allowEmpty) ? undefined : newVal
-      }
-      const transformed = _.omitBy(
-        await transformValues({
-          values: newVal,
-          type: fieldType,
-          transformFunc,
-          strict,
-          pathID: keyPathID,
-          elementsSource,
-          isTopLevel: false,
-          allowEmpty,
-        }),
-        _.isUndefined
-      )
-      return _.isEmpty(transformed) && (!_.isEmpty(newVal) || !allowEmpty) ? undefined : transformed
-    }
-    if (_.isPlainObject(newVal) && !strict) {
-      const transformed = _.omitBy(
-        await mapValuesAsync(
-          newVal ?? {},
-          (val, key) => transformValue(val, keyPathID?.createNestedID(key)),
-        ),
-        _.isUndefined,
-      )
-      return _.isEmpty(transformed) && (!allowEmpty || !_.isEmpty(newVal)) ? undefined : transformed
-    }
-    return newVal
+    const recursed = recurseIntoValue(
+      newVal,
+      transformValue,
+      strict,
+      allowEmpty,
+      true,
+      keyPathID,
+      field,
+      await field?.getType(elementsSource)
+    )
+    return removeEmptyParts(await depromise(recursed), allowEmpty)
```

Comment:
with the new mapper functions, we will always get back a single promise from `recurseIntoValue`, we don't need the `depromise` logic anymore, we can remove that function and simply change this into
```suggestion
    const recursed = await recurseIntoValue(
      newVal,
      transformValue,
      strict,
      allowEmpty,
      true,
      keyPathID,
      field,
      await field?.getType(elementsSource)
    )
    return removeEmptyParts(recursed, allowEmpty)
```